app/src/components/features/rules/RuleBuilder/Header/ActionButtons/CreateRuleButton/actions/index.js

```diff
@@ -3,10 +3,103 @@ import { actions } from "../../../../../../../../store";
 //UTILS
 import { isValidUrl } from "../../../../../../../../utils/FormattingHelper";
 import { CONSTANTS as GLOBAL_CONSTANTS } from "@requestly/requestly-core";
-// LODASH
-import { inRange } from "lodash";
+import _, { inRange } from "lodash";
 import { getFunctions, httpsCallable } from "firebase/functions";
 import { ResponseRuleResourceType } from "types/rules";
+import {
+  extractInnerTextAndAttributesFromHTMLString,
+  getHTMLNodeName,
+  validateStringForSpecificHTMLNode,
+} from "./insertScriptValidators";
+import { isFeatureCompatible } from "utils/CompatibilityUtils";
+import FEATURES from "config/constants/sub/features";
+
+/**
+ * Convert user inputs into rule schema
+ *
+ * @param {*} ruleData
+ * @returns {
+ *  success: boolean,
+ *  newRuleData?: object, // in case of successful validation and transformation
+ *
+ *  // in case of failure
+ *  error?: string,
+ *  message?: string,
+ *  pairId?: string, // id of the pair that caused the error
+ * }
+ */
+export const transformAndValidateInputsforRule = async (ruleData) => {
+  switch (ruleData.ruleType) {
+    case GLOBAL_CONSTANTS.RULE_TYPES.SCRIPT: {
+      const isCompatibleWithAttributesForScriptRule = isFeatureCompatible(FEATURES.SCRIPT_RULE_HTML_BLOCK);
+
+      if (!isCompatibleWithAttributesForScriptRule) {
+        return ruleData;
+      }
+
+      const ruleDataCopy = _.cloneDeep(ruleData);
```

Comment:
Why do we need a deep copy of rule data? Can you also add a comment here?

Reply:
RuleData is explicitly read-only (reference to value in store). Updating it directly is possible but leads to very messy code

Added a comment

Reply:
@nsrCodes I am still not sure you really need to deep clone the ruleData

Looking at this code below 👇 You are again creating a new ruleObject with new rulePairs

<img width="286" alt="image" src="https://github.com/requestly/requestly/assets/3108399/4b30404f-be16-48de-ba72-fae837ec0c29">

Just sharing my understanding of `...` operator 

<img width="294" alt="image" src="https://github.com/requestly/requestly/assets/3108399/779ae737-df85-4722-9b46-9e4e8234f1ae">

Reply:
And, Please do not resolve any conversation just after your comment, let me resolve the conversation otherwise I will miss the responses.

Reply:
This works for when we only need to change 1 level deep. 
Here the whole object immutable, down to every scrip object which is why had to deep clone